Kernel/StackWriter.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2022, Kim Nilsson <kim@wayoftao.net>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/StdLibExtraDetails.h>
+#include <AK/Types.h>
+
+namespace Kernel {
+
+class StackWriter {
+public:
+    explicit StackWriter(FlatPtr top)
+        : m_top { top }
+    {
+    }
+
+    // Decrement the stack top, reinterpret the intermediate bytes as an (uninitialized) object of type T, and return a reference to it.
+    // FIXME: This is probably technically UB, but works by tradition. See C++23 std::start_lifetime_as.
+    template<typename T>
+    [[nodiscard]] T& emplace(size_t offset = 0)
```

Comment:
`Foo::emplace` in AK constructs an object given its ctor parameters, could we use a different name for this? or make it behave like other `emplace` functions?

Reply:
I just couldn't come up with a novel name without bikeshedding for too long. A change to be more in line with traditional emplace functions (which I also tried) has two issues:
1. One of the initial target types is TrapFrame, which has its constructor explicitly deleted.
2. It doesn't interact well with the default argument, making the member function cumbersome.

Reply:
> TrapFrame deleted default ctor

not sure why that's a thing, but our current code is also skipping the ctor, I guess this only "works" because TrapFrame just so happens to be a POD type

> It doesn't interact well with the default argument

does that argument even need to be omitted anywhere?